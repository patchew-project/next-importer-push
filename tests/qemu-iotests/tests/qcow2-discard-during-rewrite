#!/usr/bin/env bash
# group: quick
#
# Test discarding (and reusing) host cluster during writing data to it.
#
# Copyright (c) 2021 Virtuozzo International GmbH.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

# creator
owner=vsementsov@virtuozzo.com

seq=`basename $0`
echo "QA output created by $seq"

status=1	# failure is the default!

_cleanup()
{
    _cleanup_test_img
}
trap "_cleanup; exit \$status" 0 1 2 3 15

# get standard environment, filters and checks
. ./../common.rc
. ./../common.filter

_supported_fmt qcow2
_supported_proto file fuse
_supported_os Linux

size=1M
_make_test_img $size

(
cat <<EOF
write -P 1 0 64K

break pwritev A
aio_write -P 2 0 64K
wait_break A

aio_discard 0 64K
EOF

# Now discard must be blocked by discard_rw_lock.
# Prior to introducing of discard_rw_lock discard
# will run and discard the cluster we are writing
# to. Give it a chance to do it.
sleep 1

# Write another cluster. Data writing should be
# blocked by discard_rw_lock (as we have pending
# writer). Prior to introducing of the lock this
# write will allocate same cluster that was
# discarded and successfully write to it.
# Give it a chance.
cat <<EOF
aio_write -P 3 128K 64K
EOF

# a chance for write to finish
sleep 1

cat <<EOF
resume A
EOF

# On good scenario wait for
# 1. finish first write
# 2. finish discard
# 3. finish second write
# On bad scenario discard and second write are finished,
# wait only for first write, which will pollute data
# of second write
sleep 1

cat <<EOF
read -P 0 0 64K
read -P 3 128K 64K
EOF
) | $QEMU_IO blkdebug::$TEST_IMG | _filter_qemu_io

# success, all done
echo "*** done"
rm -f $seq.full
status=0
