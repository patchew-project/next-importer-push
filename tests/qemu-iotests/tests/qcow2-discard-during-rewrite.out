QA output created by qcow2-discard-during-rewrite
Formatting 'TEST_DIR/t.IMGFMT', fmt=IMGFMT size=1048576
wrote 65536/65536 bytes at offset 0
64 KiB, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
blkdebug: Suspended request 'A'
blkdebug: Resuming request 'A'
discarded 65536/65536 bytes at offset 0
64 KiB, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
wrote 65536/65536 bytes at offset 0
64 KiB, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
wrote 65536/65536 bytes at offset 131072
64 KiB, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
read 65536/65536 bytes at offset 0
64 KiB, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
read 65536/65536 bytes at offset 131072
64 KiB, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
*** done
